File: src/wctcvt.py
#!/usr/bin/env python
from __future__ import with_statement
import sys

execfile('libwocecvt.py')
execfile('libwctcvt.py')
execfile('sumlim.py')

load_carter_table_file(CARTERTAB_ATB)
load_carter_table_file(CARTERTAB_TB)

def convert(sum_filename,):
  global QCODE_DQE, QCODE_PICKY, QCODE_ARCHIVE
  global USE_QUALT1, USE_QUALT2
  global use_carter_tables, qcode_mode, qualt
  global wct_filename, out_filename

  sum = []
  with open(sumfile, 'r') as sumf:
    for line in sumf:
      sum.append(line)

  del sum[0]
  begin, end = get_summary_limits(sum)
  if begin is None and end is None:
    return 1

  if begin[CDEPTH] != -1:
    pass # use available corrected depths.
  elif use_carter_tables:
    if _verbose:
      print ' bottom depths will be corrected with carter tables.'
  else:
    if _verbose:
      print ' no corrected bottom depths; no conversion.'

  wct = []
  with open(wctfile, 'r') as wctf:
    for line in wctf:
      wct.append(line)

  nstn = 0
  same_expocode_p = True

  if len(wct) == 0:
    if _verbose:
      print "%3d station/casts converted" % nstn
    return 0
  for line in wct:
    lst = []
    with open(line, 'r') as lstfile:
      for line in lstfile:
        list.append(line)

    identifiers, _ = wctun1(lst)
    if _ > 0:
      if _ == 1:
        pass
      else:
        if _verbose:
          print>>sys.stderr,'wctcvt: error in wctun1; aborting current'
        continue # to next list file

    headers, _ = wctlim(list)
    if _ > 0:
      if _ == 1:
        pass
      else:
        if _verbose:
          print>>sys.stderr,'wctcvt: error in wctlim; aborting current'
        continue # to next file
    name,units,begin,end,qualt_pos = headers

    all_items, _ = wctun2(list[6:],begin,end,identifiers['NO. RECORDS'])
    if _ > 0:
      if _ == 1:
        if _verbose:
          print>>sys.stderr,'wctcvt: warning from wctun2 for %s' % line
      else:
        if _verbose:
          print>>sys.stderr,'wctcvt: error in wctun2; aborting current'
        continue # to next list file

    previous = -999
    for pressure in all_items['DATA']['CTDPRS']:
      if pressure <= previous:
        if _verbose:
          print>>sys.stderr,'wctcvt: pressures do not increase',
          print>>sys.stderr,'monotonically; aborting current'
        previous = -999
        break # to handler
      else:
        previous = pressure
    if previous == -999: # handler
      continue # to next file

    line,expocode,_ = find_sum_line(sum,begin,end,stnnbr,castno)
    if _ != 0:
      if _ == 3:
        if _verbose:
          print>>sys.stderr,'wctcvt: cast %s not found;',
          print>>sys.stderr,'trying 1'%str(castno)
        line,expocode,_ = find_sum_line(sum,begin,end,stnnbr,1)
        if _ != 0:
          if _verbose:
            print>>sys.stderr,'wctcvt: cast 1 not found'
            print>>sys.stderr,'wctcvt: skipping stn=%s'%str(stnnbr),
            print>>sys.stderr,'cast=%s'%str(castno)
          continue # to next file
        else:
          if _verbose:
            print>>sys.stderr,'wctcvt: using cast 1',
            print>>sys.stderr,'for stn=%s'%str(stnnbr)
      elif _ == 1:
        if _verbose:
          print>>sys.stderr,'wctcvt: station %s not found'%str(stnnbr)
        continue # to next file

    if expocode.find('/') == -1:
      same_expocode_p = (expocode == identifiers['EXPOCODE'])
    else:
      same_expocode_p = (expocode[:expocode.find('/')] ==
                         identifiers['EXPOCODE']
                             [:identifiers['EXPOCODE'].find('/')])

    headers = cddcfh(line, begin, end, identifiers['DATE']['year'],
              identifiers['DATE']['month'], identifiers['DATE']['day'],
              identifiers['NO. RECORDS'], identifiers['INSTRUMENT'])
    headers[4]=(headers[4][:35]+'%7.1f'%all_items['DATA']['CTDPRS'][1]+
                headers[4][41:])
    headers[5]=(headers[5][:35]+'%7.1f'%all_items['DATA']['CTDPRS'][-1]+
                headers[5][41:])

    headers[6], header[7], h67len = cddcfu(
        map((lambda x: x[2]), headers),
        map((lambda x: x[0]), headers),
        map((lambda x: x[1]), headers))

    contents,warn,_ = cddcfd(begin, qualt_pos, identifiers['NO. RECORDS'],
                         all_items['DATA'], all_items['QUALT%1s'%qualt],
    if _ != 0:
      if _verbose:
        print>>sys.stderr,'wctcvt: error in stn=%s cast=%s'%(
                          str(stnnbr),str(castno))
    if warn:
      if _verbose:
        print>>sys.stderr,'wctcvt: warning in stn=%s cast=%s'%(
                          str(stnnbr),str(castno))
    with open(out_filename, 'a') as of:
      for line in contents:
        of.write(line)

def main():
  global qcode_mode, _qcode_mode_set, _qcode
  global qualt, _qualt_set, ocntry
  global _verbose, sum_filename, out_filename, wct_filename
  i=0
  while i<len(sys.argv):
    if sys.argv[i] in ['-h', '--help']:
      print >> sys.stderr, HELP_MSG
      exit(0)
    elif sys.argv[i] in ['-dqe', '-picky', '-archive']:
      if _verbose:
        print >> sys.stderr, 'using %s mode.' % sys.argv[i][1:]
      qcode_mode = _qcode[sys.argv[i]]
      _qcode_mode_set = True
    elif sys.argv[i] in ['-qualt1', '-qualt2']:
      if _verbose:
        print >> sys.stderr, 'using QUALT%s' % sys.argv[i][-1]
      qualt = int(sys.argv[i][-1])
      _qualt_set = True
    elif sys.argv[i] in ['-o', '--output-file']:
      i += 1
      if i >= len(sys.argv):
        print >> sys.stderr, "wctcvt: missing argument to --output-file"
        exit(1)
      out_filename = sys.argv[i]
    elif sys.argv[i] in ['-l', '--list-file']:
      i += 1
      if i >= len(sys.argv):
        print >> sys.stderr, "wctcvt: missing argument to --list-file"
        exit(1)
      wct_filename = sys.argv[i]
    elif sys.argv[i] in ['-c', '--country']:
      i += 1
      if i >= len(sys.argv):
        print >> sys.stderr, "wctcvt: missing argument to --country"
        exit(1)
      ocntry = sys.argv[i][:2]
    elif sys.argv[i] in ['-q', '--quiet']:
      _verbose = False
    else:
      sum_filename.append(sys.argv[i])
    i+=1
  if (len(sum_filename)==0 or 
      wct_filename is None or
      not _qcode_mode_set  or
      not _qualt_set       or
      out_filename is None):
    print PRGM_HDR
  if len(sum_filename) == 0:
    print ' enter input .SUM filename:'
    sum_filename.append(raw_input())
  if wct_filename is None:
    print ' enter a file containing all the .wct filenames:'
    wct_filename = raw_input()
  if not _qcode_mode_set:
    print QCODE_DESC
    while not _qcode_mode_set:
      print ' choose a mode (1-3):'
      try:
        qcode_mode = int(raw_input())
        _qcode_mode_set = (qcode_mode in range(1,4))
      except (TypeError, ValueError):
        _qcode_mode_set = False
  if not _qualt_set:
    while not _qualt_set:
      print 'Look at QUALT1 or QUALT2? (1 or 2):'
      try:
        qualt = int(raw_input)
        _qualt_set = (qualt in range(1,3))
      except (TypeError, ValueError):
        _qualt_set = False
  if out_filename is None:
    print ' enter output filename:'
    out_filename = raw_input()

File: dnewton/wocecvt_code/code/wctcvt.f
c program wctcvt reads woce format CTD data and converts it to something
c   more reasonable. (single file DCF format)
c  
c author: David Newton. Scripps Instn. of Oceanog.
C *****************
C   PORTING NOTES:
C  completely standard FORTRAN. 
C  
C    i/o units
C          li == standard input    lo == standard output
C          le == standard error output (0 on sun, 7 on HP)
C          IOA,IOB,IOC,IOL ==  .SUM file, .WCT file, data file output, list of
C                               input .WCT files.
C   On most computers you'll just have to verify that LE is correct.
C   (initialized in a data statement below declarations)
C
C   cartertab.f uses short integers internally.
C*****************
c
c changes:
c  20Apr94 dmn. first draft.
c  11Aug95 dmn. typo in external st.
c
      integer li,lo,le,ioa,iob,ioc,iol,icd
      integer lrec,nstout,i,wchqul,mxline,
     *        mxhdng,qulmod
      parameter (mxline=300,mxhdng=19)
      integer bg(mxhdng),en(mxhdng)
      logical lsteof,samexp,docart,warn
      character line*200
      character name*40,ocntry*2
c
      integer mxcnum,mxctd
      parameter (mxcnum=9,mxctd=8000)
      integer wbg(-1:mxcnum),wen(-1:mxcnum),qualps(-1:mxcnum),
     *        cyear,cmonth,cday,nobs,nparam,h67len,nast
      real x,samphz,dat(mxctd,mxcnum)
      character names(-1:mxcnum)*8,units(-1:mxcnum)*8,cexpo*14,
     *          cwhpid*5,instru*4,qual(mxctd,2)*10
      character h1*37,h2*40,h3*39,h4*41,h5*41,h6*80,h7*80
c
      character stnnbr*6,expo*14
      integer castno
c
      intrinsic real,len,index
      external sumlim,fndsum,wctlim,wctun1,wctun2,cddcfh,cddcfu,
     *         cddcfd
c
      data li/5/,lo/6/,le/7/,ioa/10/,iob/12/,ioc/14/,iol/16/
      data nstout/0/,samexp/.true./,docart/.true./
c
      write(lo,*)' program wctcvt converts WOCE format CTD data'
      write(lo,*)'   to a more workable format.    Aug 24 1995'
      write(lo,*)' enter input .SUM filename:'
      read(li,2010) name
 2010  format(a)
c
      open(ioa,file=name,status='old')
      read(ioa,*)
      call sumlim(le,ioa,bg,en)
      if(bg(19).ne.0) then
c         -- corrected depths are available.
      else if(docart) then
          write(lo,*)' bottom depths will be corrected with',
     *               ' carter tables.'
      else
          write(lo,*)' no corrected bottom depths. no conversion.'
      endif
c
c -- ask for filename that contains the *.WCT filesnames.
      write(lo,*)' enter a file containing all the .wct filenames:'
      read(li,2010) name
      open(iol,file=name,status='old')
c
c -- ask which way to treat quality codes
      qulmod = 3
      write(lo,*)' quality code mode:'
      write(lo,*)'  1. DQE mode. All data considered good.'
      write(lo,*)'  2. Picky mode. Only good data will be allowed.'
      write(lo,*)'  3. Archive mode. Uncertain codes preserved.'
      write(lo,*)' choose a mode (1-3):'
      read(li,*) i
      
      if(i.ge.1 .and. i.le.3) then
          qulmod = i
      else
c         -- defaults to archive mode
          write(lo,*)' using Archive mode for quality codes.'
          qulmod = 3
      endif
c
c -- ask if want to look at QUALT1 or QUALT2.
      write(lo,*)' Look at QUALT1 or QUALT2? (1 or 2):'
      read(li,*) i
      wchqul = 1
      if(i .eq. 1 .or. i .eq. 2) then
          wchqul = i
      else 
          write(lo,*)' incorrect response. using QUALT1.'
      endif
c
c -- ask for originators country. 
      write(lo,*)' enter 2 char originator''s country code:'
      read(li,2010) ocntry
c
c -- ask for output filename.
      write(lo,*)' enter output filename:'
      read(li,2010) name
      open(ioc,file=name,status='new')
c
c -- loop through the list file.
      lrec = 0
      lsteof = .false.
 40   continue
      lrec = lrec + 1
      read(iol,2010,iostat=icd) name
      if(icd .ne. 0) then
          if(icd .lt. 0) then
c             -- hit eof on .lst file. go to summary.
              lsteof = .true.
              go to 400
          else
            write(le,*)'wctcvt: read err on list of files irec,iostat=',
     *                   lrec,icd
              stop 99
          endif
      endif
c
c -- open woce ctd data file.
      open(iob,file=name,status='old',iostat=icd)
c
c -- read the header info in the woce ctd file.
      call wctun1(iob,le,cexpo,cwhpid,cyear,cmonth,cday,stnnbr,
     *            castno,nobs,instru,samphz,icd)
      if(icd .ne. 0) then
          if(icd .eq. 1) then
c             -- just a warning. try to continue.
          else
              write(le,*)'wctcvt: err rtn frm wctun1. skipping file=',
     *                    name
              go to 400
          endif
      endif
c
c -- find data limits in the woce ctd data file.
      call wctlim(iob,le,wbg,wen,qualps,nast,nparam,names,units,icd) 
      if(icd .ne. 0) then
          write(le,*)'wctcvt: err rtn frm wctlim. trying to figure out',
     *               'where things are. skipping file=',name
          go to 400
      endif
c
c -- read in a whole station of data.
      warn = .false.
      call wctun2(iob,le,wbg,wen,nast,nparam,nobs,dat,qual,warn,icd)
      if(icd .ne. 0) then
          write(le,*)'wctcvt: err rtn frm wctun2. trying to read data.',
     *               'skipping file=',name
          go to 400
      endif
      if(warn) then
          write(le,*)'wctcvt: warning frm wctun2 in file=',name
      endif
c
c -- get to here to process a full station of data just read. 
 200  continue
c     -- test for out of order and repeated pressures.
      x = dat(1,1)
      do 210 i=2,nobs
         if(dat(i,1) .gt. x) then
             x = dat(i,1)
         else
c            -- data in not pressure ascending.
             write(le,*)'wctcvt: pressures non-ascending or repeated.',
     *                  'skipping file=',name,' press=',x
             go to 400
         endif
 210  continue
c
c    -- find correct .sum line.
      call fndsum(ioa,le,bg,en,stnnbr,castno,line,expo,icd)
      if(icd .ne. 0) then
          if(icd .eq. 3) then
              write(le,*)'wctcvt: cast',castno,
     *                   ' not found. will try cast 1'
              call fndsum(ioa,le,bg,en,stnnbr,1,line,expo,icd)
              if(icd .ne. 0) then
                  write(le,*)'wctcvt: cast 1 not found either.'
                  write(le,2060) stnnbr,castno
 2060             format('wctcvt: stn= ',a,' cast=',i4,' skipping.')
                  go to 400
              else
c                 -- okay, but had to use cast 1.
                  write(le,*)'wctcvt: will use .SUM info from',
     *                       ' cast 1.'
              endif
          else if(icd .eq. 1) then
              write(le,*)'wctcvt: station not found.'
              write(le,2060) stnnbr,castno
              go to 400
          endif
      endif
c
c -- check if expocode from .SUM file matches that on the .WCT file.
      i = index(expo,'/') - 1
      if(i .eq. -1) i = len(expo)
      if(expo(1:i) .ne. cexpo(1:i)) then
          samexp = .false.
      endif
c
c -- code the first 5 DCF headers.
      call cddcfh(line,le,bg,en,docart,cyear,cmonth,cday,nobs,ocntry,
     *            instru,h1,h2,h3,h4,h5)
c     -- note: minprs + maxprs still need to be coded into 4th + 5th headers.
      write(h4(35:41),'(f7.1)')dat(1,1)
      write(h5(35:41),'(f7.1)')dat(nobs,1)
c
c -- code the name + units headers.
      call cddcfu(le,wbg,names,units,h6,h7,h67len)
c
c -- output the completed header lines.
      write(ioc,2010) h1
      write(ioc,2010) h2
      write(ioc,2010) h3
      write(ioc,2010) h4
      write(ioc,2010) h5
      write(ioc,2010) h6(1:h67len)
      write(ioc,2010) h7(1:h67len)
c
c -- call routines to form and write out the data lines.
c
      warn = .false.
      call cddcfd(le,ioc,wbg,qualps,nobs,dat,qual(1,wchqul),
     *            qulmod,warn,icd)
      if(icd .ne. 0) then
          write(le,*)'wctcvt: errs above from stn/cast= ',stnnbr,'/',
     *                castno
          stop 99
      endif
      if(warn) then
         write(le,*)'wctcvt: warnings above from stn/cast= ',stnnbr,'/',
     *                castno
      endif
c
      nstout = nstout + 1
c -- close the wct data file.
      close(iob)
 400  if(lsteof) then
c         -- done with .LST file. done with everything.  normal exit.
          close(ioa)
          close(iob)
          close(ioc)
          if(.not. samexp) then
              write(le,*)' ** expocode did not match in .SUM +.WCT'
          endif
          write(le,*)' station/casts converted=',nstout
          stop
      else
          go to 40
      endif
c
      end
